### ISSUE TITLE
Different results got using different batch sizes (cannot fully fit into GPU memory therefore evaluating different batch each time)

### ISSUE THREAD

**Issue by ZisIsNotZis** [tensorflow/tensorflow#23182]
**System information**
- Have I written custom code: **Yes**
- OS Platform and Distribution: **Linux Ubuntu 18.04.1** 
- Mobile device: N/A
- TensorFlow installed from: **binary, tensorflow-gpu from anaconda**
- TensorFlow version: **b'unknown', 1.11.0**
- Python version: **3.6.6**
- Bazel version: N/A
- GCC/Compiler version: N/A
- CUDA/cuDNN version: **10.0, Driver Version: 410.66**
- GPU model and memory: **TITAN Xp, 12195MiB**

**Describe the current behavior**
I'm trying to batch evaluate MLP (with same structure but different parameter) on GPU. Since the whole batch is too big to fit into GPU memory, I have to create smaller batches from the whole batch. Since the size of whole batch is not divisible by the size of little batch, the last little batch will be smaller than usual. It turns out the results in the last small batch is way smaller than previous batches.

**Describe the expected behavior**
The result should be the same since batch size should not affect the result itself.

**Exact command to reproduce**
Put any of the following examples in a `jupyter` cell, and run it
Put any of the following examples in `xxx.py`, and run `python xxx.py` should also be fine

**Code to reproduce the issue**
A small example:

    import tensorflow as tf
    import numpy as np
    with tf.Session():
        x = tf.Variable(np.random.rand(43, 60000).astype('f2'))
        w = tf.Variable(np.ones((58624, 43), 'f2'))
        b = tf.Variable(np.ones((58624, 1), 'f2'))
        v1 = w @ x + b
        w = tf.Variable(np.ones((5376, 43), 'f2'))
        b = tf.Variable(np.ones((5376, 1), 'f2'))
        v2 = w @ x + b
        tf.global_variables_initializer().run()
        print(v1.eval(), v2.eval()) # they should have the same value while they dont

The actual example:

    import tensorflow as tf
    import numpy as np
    with tf.Session():
        x = tf.Variable(np.ones((43, 60000), 'f2'))
        y = tf.Variable(np.ones((8, 60000), 'f2'))
        w = tf.Variable(np.ones((916* 64, 43), 'f2'))
        b = tf.Variable(np.ones((916* 64, 1), 'f2'))
        v = tf.Variable(np.ones((916, 8, 64), 'f2'))
        c = tf.Variable(np.ones((916, 8, 1), 'f2'))
        r = tf.linalg.norm(tf.cast(v@tf.reshape(tf.tanh(w@x+b),(916,64,60000))+c,'float32'),axis=(1,2))# prevent overflow
        w = tf.Variable(np.ones((84* 64, 43), 'f2'))
        b = tf.Variable(np.ones((84* 64, 1), 'f2'))
        v = tf.Variable(np.ones((84, 8, 64), 'f2'))
        c = tf.Variable(np.ones((84, 8, 1), 'f2'))
        R = tf.linalg.norm(tf.cast(v@tf.reshape(tf.tanh(w@x+b),(84,64,60000))+c,'float32'),axis=(1,2)) # prevent overflow
        tf.global_variables_initializer().run()
        print(r.eval()[::100], R.eval()[::10]) # summary, they should have the same value while they dont

**Other info / logs**
I created a stack overflow question (https://stackoverflow.com/questions/52921884/tensorflow-result-inconsistent-across-batch-size) but since no body solved the problem, I submitted an issue here. It seems that this problem does not occur with small batches, it sometimes is not stably reproducible

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Exact command to reproduce

**Comment by ZisIsNotZis** [comment]
Sorry for the delay (12h time zone difference), updated accordingly

**Comment by wt-huang** [comment]
@ZisIsNotZis I was able to run the first code snippet in GPU and received consistent results across all batches. Didn't observe different result in last batch as you mentioned.

Also, I noticed that the first code runs very slowly and consumes a lot of memory, i.e. close to 8G, which may be the reason for abnormal behavior. You may need to do some optimization there.

**Comment by ZisIsNotZis** [comment]
That's wired. I just updated everything to latest (i.e. `conda update --all`), including ubuntu drivers (i.e. `apt update` and `ubuntu-drivers autoinstall`) and the result is still inconsistent. For first code, it gives the result:

    [[23.03 22.25 20.97 ... 22.72 25.58 20.19]
     [23.03 22.25 20.97 ... 22.72 25.58 20.19]
     [23.03 22.25 20.97 ... 22.72 25.58 20.19]
     ...
     [23.03 22.25 20.97 ... 22.72 25.58 20.19]
     [23.03 22.25 20.97 ... 22.72 25.58 20.19]
     [23.03 22.25 20.97 ... 22.72 25.58 20.19]] [[24.03 23.25 21.97 ... 23.72 26.58 21.19]
     [24.03 23.25 21.97 ... 23.72 26.58 21.19]
     [24.03 23.25 21.97 ... 23.72 26.58 21.19]
     ...
     [24.03 23.25 21.97 ... 23.72 26.58 21.19]
     [24.03 23.25 21.97 ... 23.72 26.58 21.19]
     [24.03 23.25 21.97 ... 23.72 26.58 21.19]]

where `v2.eval()` is apparently off `v1.eval()` by 1. The second code gives:

    [1906641.6 1906641.6 1906641.6 1906641.6 1906641.6 1906641.6 1906641.6
     1906641.6 1906641.6 1906641.6] [45033.324 45033.324 45033.324 45033.324 45033.324 45033.324 45033.324
     45033.324 45033.324]

which is also inconsistent. I'm wondering if that is a anaconda-specific problem since I didn't do the standard `pip install tensorflow-gpu`. It might also be a gpu-memory-size problem since you might have an even better GPU.

For optimizing part, I thought I would get higher efficiency if put larger batch (i.e. larger matrix multiplication) on GPU instead of smaller batch multiple times. Is that correct? Do I have to split matrix multiplication manually or there are some function in Tensorflow that can help me with that?

Thank you!

**Comment by ZisIsNotZis** [comment]
I just tested with the standard pip-installed `tensorflow-gpu`, it has the same behavior.

- TensorFlow installed from: **pip, tensorflow-gpu**
- TensorFlow version: **v1.12.0-0-ga6d8ffae09, 1.12.0**
- Python version: **3.6.7**
- CUDA/cuDNN version: **10.0, Driver Version: 410.73**

**Comment by tensorflowbutler** [comment]
Nagging Assignee @wt-huang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ZisIsNotZis** [comment]
Hi, I'm sorry but I'm wondering what is the status of this issue? I'm doing some research about neural network optimization. It's more about understanding what it looks like etc. and I kind of cannot go on if the computation result might be wrong.

Thank you!

**Comment by jerry-rubikloud** [comment]
I have similar issues. Basically, it happens also if you are using CPU only. 
If I feed the same input data for training to the evaluation to the estimator, the loss reported is consistent between training and evaluation but if I change the batch size during the evaluation, the loss is very off. It looks like a bug in the way the avg loss is calculated.

**Comment by gadagashwini** [comment]
@ZisIsNotZis I tried with latest stable version of Tensorflow 1.14.0. I got the expected result as below
```
[1906641.6 1906641.6 1906641.6 1906641.6 1906641.6 1906641.6 1906641.6
 1906641.6 1906641.6 1906641.6] [45033.324 45033.324 45033.324 45033.324 45033.324 45033.324 45033.324
 45033.324 45033.324]
```
Can you check with latest stable version. Let us know. Thanks!

**Comment by gadagashwini** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
